File: client-nodejs/src/tsd-counter.ts
/*
 * Copyright 2014 Groupon.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/* istanbul ignore next */ //ignores inheritance snippet emitted by the TS compiler

import tsdDef = require("tsdDef");
import tsd = require("tsd-metrics-client");
import sample = require("./tsd-metric-sample");

//aliases
import TsdMetricSample = sample.TsdMetricSample;

/**
 * Class for counter. Instances are initialized to zero on creation. The zero-value sample is recorded when the
 * [TsdMetrics]{@linkcode module:tsd-metrics-client~TsdMetrics} instance is closed if no other actions are taken on the
 * <code>Counter</code>.
 *
 * Modifying the <code>Counter</code> instance's value modifies the single sample value. When the associated
 * [TsdMetrics]{@linkcode module:tsd-metrics-client~TsdMetrics} instance is closed whatever value the sample has is
 * recorded. To create another sample you create a new <code>Counter</code> instance with the same name.
 *
 * Each Counter instance is bound to a [TsdMetrics]{@linkcode module:tsd-metrics-client~TsdMetrics} instance. After the
 * [TsdMetrics]{@linkcode module:tsd-metrics-client~TsdMetrics} instance is closed any modifications to the
 * <code>Counter</code> instances bound to that [TsdMetrics]{@linkcode module:tsd-metrics-client~TsdMetrics} instance
 * will be ignored.
 *
 * @class
 * @extends MetricSample
 * @alias Counter
 * @author Mohammed Kamel (mkamel at groupon dot com)
 */
export class TsdCounter extends TsdMetricSample implements tsdDef.Counter {

    /**
     * Constructor.
     *
     * @param {MetricsStateObject} _metricsStateObject Object holding state of the parent metrics object.
     * @ignore
     */
    constructor(private _name:string, private _metricsStateObject:tsd.MetricsStateObject) {
        super(0, undefined);
    }

    /**
     * Increment the counter sample by the specified value.
     *
     * @method
     * @param {number} value The value to increment the counter by. Default = 1
     * @emits 'error' if the metrics object is closed
     */
    public increment(value:number = 1):void {
        if (this._metricsStateObject.assertIsOpen("Cannot increment/decrement counter '" + this._name + "'")) {
            (<any>this)._value = (<any>this)._value + value;
        }
    }

    /**
     * Decrement the counter sample by the specified value.
     *
     * @method
     * @param {number} value The value to decrement the counter by. Default = 1
     * @emits 'error' if the metrics object is closed
     */
    public decrement(value:number = 1):void {
        this.increment(-1 * value);
    }
}

File: client-java/src/main/java/com/arpnetworking/metrics/impl/TsdCounter.java
/**
 * Copyright 2014 Groupon.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.arpnetworking.metrics.impl;

import com.arpnetworking.metrics.Counter;
import com.arpnetworking.metrics.Quantity;
import com.arpnetworking.metrics.Unit;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Implementation of <code>Counter</code>.  This class is thread safe.
 *
 * @author Ville Koskela (vkoskela at groupon dot com)
 */
final class TsdCounter implements Counter, Quantity {
    /**
     * Package private constructor. All <code>TsdCounter</code> instances should
     * be created through the <code>TsdMetrics</code> instance.
     * 
     * @param name The name of the counter.
     * @param isOpen Reference to state of containing <code>TsdMetrics</code>
     * instance. This is provided as a separate parameter to avoid creating a
     * cyclical dependency between <code>TsdMetrics</code> and 
     * <code>TsdCounter</code> which could cause garbage collection delays.
     */
    /* package private */static TsdCounter newInstance(final String name, final AtomicBoolean isOpen) {
        return new TsdCounter(name, isOpen, DEFAULT_LOGGER);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void increment() {
        increment(1L);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void decrement() {
        increment(-1L);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void increment(final long value) {
        assertIsOpen();
        _value.addAndGet(value);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void decrement(final long value) {
        increment(-1L * value);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String toString() {
        return String.format(
                "TsdCounter{id=%s, Name=%s, Value=%s, IsOpen=%s}",
                Integer.toHexString(System.identityHashCode(this)),
                _name,
                _value,
                _isOpen);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public Number getValue() {
        return Long.valueOf(_value.get());
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public Unit getUnit() {
        return null;
    }

    private boolean assertIsOpen() {
        final boolean isOpen = _isOpen.get();
        if (!isOpen) {
            // This is in place of an exception
            _logger.warn(String.format("Counter manipulated after metrics instance closed; counter=%s", this));
        }
        return isOpen;
    }

    // NOTE: Package private for testing.
    /* package private */TsdCounter(final String name, final AtomicBoolean isOpen, final Logger logger) {
        _name = name;
        _isOpen = isOpen;
        _logger = logger;
        _value = new AtomicLong(0L);
    }

    private final String _name;
    private final AtomicBoolean _isOpen;
    private final Logger _logger;
    private final AtomicLong _value;

    private static final Logger DEFAULT_LOGGER = LoggerFactory.getLogger(TsdCounter.class);
}
